meta-balena-common/recipes-support/resin-init/resin-init-flasher/resin-init-flasher

```diff
@@ -192,39 +192,35 @@ IMAGE_FILE_SIZE=$(wc -c /opt/"$BALENA_IMAGE" | awk '{print $1}')
-    if [ "$SECUREBOOT_VAR" -eq "1" ]; then
+    SECUREBOOT_VAR=$(efivar -p -n 8be4df61-93ca-11d2-aa0d-00e098032b8c-SecureBoot \
+        | awk 'NR==1, $1 == "Value:" {next}; NF {print $2}')
+    SETUPMODE_VAR=$(efivar -p -n 8be4df61-93ca-11d2-aa0d-00e098032b8c-SetupMode \
+        | awk 'NR==1, $1 == "Value:" {next}; NF {print $2}')
+
+    if [ "$SECUREBOOT_VAR" = "1" ]; then
         LUKS=1
-    fi
-    SETUPMODE_VAR=$(efivar -p -n 8be4df61-93ca-11d2-aa0d-00e098032b8c-SetupMode | awk 'NR==1, $1 == "Value:" {next}; NF {print $2}')
-    if [ "$SETUPMODE_VAR" -eq "1" ]; then
-        inform "Secure boot setup mode detected - programming keys"
-        if ! mount | grep -q "type efivarfs"; then
-            mount -t efivarfs efivarfs "${EFIVARS_MOUNTDIR}"
-        fi
-        # Leave PK last
-        for e in db KEK PK; do
-            # Remove immutable attribute
-            chattr -i ${EFIVARS_MOUNTDIR}/${e}* > /dev/null || true
-            if [ -f "${EXTERNAL_DEVICE_BOOT_PART_MOUNTPOINT}/balena-keys/${e}.auth" ]; then
-                /usr/bin/efi-updatevar -f "${EXTERNAL_DEVICE_BOOT_PART_MOUNTPOINT}/balena-keys/${e}.auth" "${e}"
-            fi
-        done
-        # Re-read setup mode configuration
-        SETUPMODE_VAR=$(efivar -p -n 8be4df61-93ca-11d2-aa0d-00e098032b8c-SetupMode | awk 'NR==1, $1 == "Value:" {next}; NF {print $2}')
-        SECUREBOOT_VAR=$(efivar -p -n 8be4df61-93ca-11d2-aa0d-00e098032b8c-SecureBoot | awk 'NR==1, $1 == "Value:" {next}; NF {print $2}')
-        if [ "$SETUPMODE_VAR" -eq "1" ]; then
-            # Setting up keys hasn't disabled setup mode, try a reboot into flasher
-            if [ "$LUKS" -ne "1" ]; then
+        if [ "$SETUPMODE_VAR" = "1" ]; then
+            inform "Secure boot setup mode detected - programming keys"
+            # Enroll PK last, as it should disable setup mode
+            for e in db KEK PK; do
+                # Remove immutable attribute
+                chattr -i ${EFIVARS_MOUNTDIR}/${e}* > /dev/null || true
+                if [ -f "${EXTERNAL_DEVICE_BOOT_PART_MOUNTPOINT}/balena-keys/${e}.auth" ]; then
+                    /usr/bin/efi-updatevar -f "${EXTERNAL_DEVICE_BOOT_PART_MOUNTPOINT}/balena-keys/${e}.auth" "${e}"
+                fi
+            done
+            # Check that we're in user mode
+            SETUPMODE_VAR=$(efivar -p -n 8be4df61-93ca-11d2-aa0d-00e098032b8c-SetupMode \
+                | awk 'NR==1, $1 == "Value:" {next}; NF {print $2}')
+            if [ "$SETUPMODE_VAR" = "1" ]; then
+                # Setting up keys hasn't disabled setup mode, try a reboot into flasher
                 inform "Rebooting flasher after secure mode setup to boot in secure boot mode."
                 # Make sure to reboot into the installer
                 bootcurrent=$(efibootmgr | grep BootCurrent | awk '{print $2}')
                 efibootmgr -n "${bootcurrent}"
                 sync
                 reboot -f
             fi
-        else
-            LUKS=1
```

Comment:
This is already set

Reply:
Not always - if `SECUREBOOT_VAR` is not set but `SETUPMODE_VAR` is it won't be set, and we want to set it after setting up the keys.

Reply:
@jakogut was this refactoring really needed when the PR had been open for months and was just reviewed and tested a couple of days ago?
What are the improvements this change brings?

Reply:
It is with the changes you introduce in this PR.